{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: ggplot2\n",
      "\n",
      "Loading required package: dplyr\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: edgeR\n",
      "\n",
      "Loading required package: limma\n",
      "\n",
      "Loading required package: plotrix\n",
      "\n",
      "Loading required package: gtools\n",
      "\n",
      "Loading required package: gplots\n",
      "\n",
      "\n",
      "Attaching package: ‘gplots’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:plotrix’:\n",
      "\n",
      "    plotCI\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    lowess\n",
      "\n",
      "\n",
      "Loading required package: eulerr\n",
      "\n",
      "Registered S3 method overwritten by 'eulerr':\n",
      "  method    from  \n",
      "  plot.venn gplots\n",
      "\n",
      "\n",
      "Attaching package: ‘eulerr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:gplots’:\n",
      "\n",
      "    venn\n",
      "\n",
      "\n",
      "Loading required package: vioplot\n",
      "\n",
      "Loading required package: sm\n",
      "\n",
      "Warning message:\n",
      "“no DISPLAY variable so Tk is not available”\n",
      "Package 'sm', version 2.2-5.6: type help(sm) for summary information\n",
      "\n",
      "Loading required package: zoo\n",
      "\n",
      "\n",
      "Attaching package: ‘zoo’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    as.Date, as.Date.numeric\n",
      "\n",
      "\n",
      "Loading required package: VennDiagram\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Loading required package: futile.logger\n",
      "\n",
      "\n",
      "Attaching package: ‘futile.logger’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:gtools’:\n",
      "\n",
      "    scat\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# pchic_pipeline.R\n",
    "# Author: Michael Song\n",
    "# Last modified: 2019-01-15\n",
    "# This script takes sets of PCHi-C interactions, ATAC-seq peaks, and RNA-seq data, performs integrative analysis, and generates tables/figures.\n",
    "#\n",
    "\n",
    "\n",
    "# Load dependencies.\n",
    "require(ggplot2)\n",
    "require(dplyr)\n",
    "require(edgeR)\n",
    "require(plotrix)\n",
    "require(gtools)\n",
    "require(gplots)\n",
    "require(eulerr)\n",
    "require(vioplot)\n",
    "require(VennDiagram)\n",
    "\n",
    "\n",
    "# Set up environment and load settings ------------------------------------\n",
    "\n",
    "\n",
    "# Clear workspace before running script.\n",
    "rm(list=ls())\n",
    "\n",
    "# Turn off scientific notation for writing output.\n",
    "options(scipen=999)\n",
    "\n",
    "# Convert warnings into errors.\n",
    "options(warn=1)\n",
    "\n",
    "\n",
    "# Set home directory.\n",
    "home.dir <- \"/shen/shenlabstore3/neng/20200219_MGENMOT\"\n",
    "#start.index <- cli.args[2]\n",
    "#end.index <- cli.args[3]\n",
    "#disease <- cli.args[4]\n",
    "\n",
    "# Home directory should be structured as such:\n",
    "# ~/scripts           - contains all scripts used in the analysis\n",
    "# ~/scripts/config.R  - configuration file\n",
    "# ~/pipeline          - contains all raw input files used in the analysis\n",
    "# ~/figures           - default output folder\n",
    "# ~/figures/saved     - contains saved intermediate files used in the analysis\n",
    "# ~/resources         - contains supplemmentary/reference files used in the analysis\n",
    "\n",
    "\n",
    "# Set up analysis environment ---------------------------------------------\n",
    "\n",
    "\n",
    "# Source files should be in ~/scripts.\n",
    "scripts.dir <- paste0(home.dir, \"/scripts\")\n",
    "source(paste0(scripts.dir, \"/interaction_analysis.R\"))\n",
    "source(paste0(scripts.dir, \"/prepare_plots.R\"))\n",
    "source(paste0(scripts.dir, \"/preprocess_data.R\"))\n",
    "source(paste0(scripts.dir, \"/read_data.R\"))\n",
    "source(paste0(scripts.dir, \"/utilities.R\"))\n",
    "\n",
    "# Load configuration file.\n",
    "source(paste0(scripts.dir, \"/config.R\"))\n",
    "\n",
    "# Create output directories.\n",
    "output.dir <- paste0(home.dir, \"/figures\")\n",
    "dir.list <- c(\"\", \n",
    "              \"/atac_seq_results\",\n",
    "              \"/feature_results\", \"/feature_results/snps\", \"/feature_results/vista\",\n",
    "              \"/interaction_results\", \"/interaction_results/general\", \"/interaction_results/hubs\", \"/interaction_results/other\", \"/interaction_results/specificity\",\n",
    "              \"/rna_seq_results\",\n",
    "              \"/saved\", \n",
    "              \"/visualization\")\n",
    "for (dir in dir.list)\n",
    "  if (!file.exists(paste0(output.dir, \"/\", run.id, dir)))\n",
    "    dir.create(paste0(output.dir, \"/\", run.id, dir))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                     V1                V2\n",
      "1 MG_rep1.genes.results         Microglia\n",
      "2 MG_rep2.genes.results         Microglia\n",
      "3 EN_rep1.genes.results ExcitatoryNeurons\n",
      "4 EN_rep2.genes.results ExcitatoryNeurons\n",
      "5 MN_rep1.genes.results      MotorNeurons\n",
      "6 MN_rep2.genes.results      MotorNeurons\n",
      "                     V1                V2\n",
      "1 MG_rep1.genes.results         Microglia\n",
      "2 MG_rep2.genes.results         Microglia\n",
      "3 EN_rep1.genes.results ExcitatoryNeurons\n",
      "4 EN_rep2.genes.results ExcitatoryNeurons\n",
      "5 MN_rep1.genes.results      MotorNeurons\n",
      "6 MN_rep2.genes.results      MotorNeurons\n",
      "                     V1                V2\n",
      "1 MG_rep1.genes.results         Microglia\n",
      "2 MG_rep2.genes.results         Microglia\n",
      "3 EN_rep1.genes.results ExcitatoryNeurons\n",
      "4 EN_rep2.genes.results ExcitatoryNeurons\n",
      "5 MN_rep1.genes.results      MotorNeurons\n",
      "6 MN_rep2.genes.results      MotorNeurons\n",
      "[1] \"Read in 57783 unique gene IDs on main chromosomes.\"\n",
      "[1] \"After filtering for allowed gene types, retained 57783 unique gene IDs.\"\n",
      "[1] \"Read in 57783 unique gene IDs on main chromosomes.\"\n",
      "[1] \"After filtering for allowed gene types, retained 34401 unique gene IDs.\"\n",
      "[1] \"Read in 57783 unique gene IDs on main chromosomes.\"\n",
      "[1] \"After filtering for allowed gene types, retained 23382 unique gene IDs.\"\n"
     ]
    }
   ],
   "source": [
    "# Read in interactions.\n",
    "#interactions.raw <- read.interactions(interactions.list, paste0(home.dir, \"/pipeline\"))\n",
    "#save(interactions.raw, file=paste0(output.dir, \"/\", run.id, \"/saved/interactions.raw.Rdata\"))\n",
    "load(file=paste0(output.dir, \"/\", run.id, \"/saved/interactions.raw.Rdata\"))\n",
    "\n",
    "# Read in ATAC-seq peaks.\n",
    "#atac.seq.peaks <- read.atac.seq.data(atac.seq.peak.list, paste0(home.dir, \"/pipeline\"), allowed.chrs, max.atac.seq.peaks)\n",
    "#save(atac.seq.peaks, file=paste0(output.dir, \"/\", run.id, \"/saved/atac.seq.peaks.Rdata\"))\n",
    "load(file=paste0(output.dir, \"/\", run.id, \"/saved/atac.seq.peaks.Rdata\"))\n",
    "\n",
    "# Read in features.\n",
    "#features <- read.features(features.list, paste0(home.dir, \"/pipeline\"))\n",
    "#save(features, file=paste0(output.dir, \"/\", run.id, \"/saved/features.Rdata\"))\n",
    "#load(file=paste0(output.dir, \"/\", run.id, \"/saved/features.Rdata\"))\n",
    "\n",
    "# Read in RNA-seq transcript quantification results.\n",
    "transcript.lengths.individual <- read.rna.seq.data(rna.seq.output.list.replicates, paste0(home.dir, \"/pipeline\"), \"length\")\n",
    "transcript.counts.individual <- read.rna.seq.data(rna.seq.output.list.replicates, paste0(home.dir, \"/pipeline\"), \"expected_count\")\n",
    "transcript.tpm.individual <- read.rna.seq.data(rna.seq.output.list.replicates, paste0(home.dir, \"/pipeline\"), \"TPM\")\n",
    "save(transcript.lengths.individual, file=paste0(output.dir, \"/\", run.id, \"/saved/transcript.lengths.individual.Rdata\"))\n",
    "save(transcript.counts.individual, file=paste0(output.dir, \"/\", run.id, \"/saved/transcript.counts.individual.Rdata\"))\n",
    "save(transcript.tpm.individual, file=paste0(output.dir, \"/\", run.id, \"/saved/transcript.tpm.individual.Rdata\"))\n",
    "load(file=paste0(output.dir, \"/\", run.id, \"/saved/transcript.lengths.individual.Rdata\"))\n",
    "load(file=paste0(output.dir, \"/\", run.id, \"/saved/transcript.counts.individual.Rdata\"))\n",
    "load(file=paste0(output.dir, \"/\", run.id, \"/saved/transcript.tpm.individual.Rdata\"))\n",
    "\n",
    "# Get cell types.\n",
    "if (exists(\"interactions.raw\")) {\n",
    "  cell.types <- names(interactions.raw)\n",
    "} else {\n",
    "  cell.types <- names(atac.seq.peaks)\n",
    "}\n",
    "\n",
    "# Get lists of promoter intervals.\n",
    "allowed.types <- c(\"protein_coding\", \"miRNA\", \"miscRNA\", \"rRNA\", \"snoRNA\", \"snRNA\", \"lincRNA\")\n",
    "promoters.all <- read.promoters(paste0(home.dir, \"/resources/\", annotation.file), allowed.chrs, tss.upstream, tss.downstream, NULL)\n",
    "promoters.strict <- read.promoters(paste0(home.dir, \"/resources/\", annotation.file), allowed.chrs, tss.upstream, tss.downstream, allowed.types)\n",
    "other.types <- unique(promoters.all$gene_type)[!(unique(promoters.all$gene_type) %in% allowed.types)]\n",
    "promoters.other <- read.promoters(paste0(home.dir, \"/resources/\", annotation.file), allowed.chrs, tss.upstream, tss.downstream, other.types)\n",
    "\n",
    "# Read in exonic and intronic intervals.\n",
    "exons <- read.table(paste0(home.dir, \"/resources/\", exons.file), sep=\"\\t\", header=F, stringsAsFactors=F, colClasses=c(\"character\", \"integer\", \"integer\", \"character\", \"numeric\", \"character\"))\n",
    "colnames(exons) <- c(\"chrom\", \"chromStart\", \"chromEnd\", \"name\", \"score\", \"strand\")                      \n",
    "introns <- read.table(paste0(home.dir, \"/resources/\", introns.file), sep=\"\\t\", header=F, stringsAsFactors=F, colClasses=c(\"character\", \"integer\", \"integer\", \"character\", \"numeric\", \"character\"))\n",
    "colnames(introns) <- c(\"chrom\", \"chromStart\", \"chromEnd\", \"name\", \"score\", \"strand\")     \n",
    "\n",
    "# Read in RMAP and BAITMAP CHiCAGO design files.\n",
    "rmap <- read.table(paste0(home.dir, \"/resources/\", rmap.file), sep=\"\\t\", stringsAsFactors=F, header=F)\n",
    "colnames(rmap) <- c(\"chrom\", \"chromStart\", \"chromEnd\", \"rmap_ID\")\n",
    "if (all(grepl(\"chr\", rmap$chrom) == FALSE)) {\n",
    "  rmap$chrom <- paste0(\"chr\", rmap$chrom)\n",
    "}\n",
    "baitmap <- read.table(paste0(home.dir, \"/resources/\", baitmap.file), sep=\"\\t\", stringsAsFactors=F, header=F)\n",
    "colnames(baitmap) <- c(\"chrom\", \"chromStart\", \"chromEnd\", \"rmap_ID\", \"baitmap_ID\")\n",
    "if (all(grepl(\"chr\", baitmap$chrom) == FALSE)) {\n",
    "  baitmap$chrom <- paste0(\"chr\", baitmap$chrom)\n",
    "}\n",
    "\n",
    "# Read in chromosome sizes.\n",
    "chrom.sizes <- read.table(paste0(home.dir, \"/resources/\", chrom.sizes.file), sep=\"\\t\", header=F, stringsAsFactors=F, colClasses=c(\"character\", \"integer\"))\n",
    "\n",
    "# Read in TAD data.\n",
    "#tads <- read.table(file=paste0(home.dir, \"/resources/\", tads.file), sep=\"\\t\", header=F, stringsAsFactors=F)\n",
    "\n",
    "# Read in positive vista elements.\n",
    "#vista.data <- read.table(file=paste0(home.dir, \"/resources/\", vista.file), sep=\"\\t\", header=F, stringsAsFactors=F)\n",
    "#colnames(vista.data) <- c(\"chrom\", \"chromStart\", \"chromEnd\", \"vista_ID\", \"annotation\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Preprocess expression data ----------------------------------------------\n",
    "\n",
    "# Calculate mean TPM or TMM-normalized FPKM expression values for each gene and cell type.\n",
    "tpm.data.individual <- summarizeExpressionResults(transcript.tpm.individual, type=\"TPM\")\n",
    "rpkm.data.individual <- summarizeExpressionResults(transcript.counts.individual, transcript.lengths.individual, type=\"RPKM\")\n",
    "\n",
    "# Write results to file.\n",
    "write.table(tpm.data.individual, file=paste0(output.dir, \"/\", run.id, \"/rna_seq_results/tpm.data.individual.txt\"), sep=\"\\t\", row.names=T, col.names=T, quote=F)\n",
    "write.table(rpkm.data.individual, file=paste0(output.dir, \"/\", run.id, \"/rna_seq_results/rpkm.data.individual.txt\"), sep=\"\\t\", row.names=T, col.names=T, quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save results.\n",
    "#save(interactions.sig.res, file=paste0(output.dir, \"/\", run.id, \"/saved/interactions.sig.res.Rdata\"))\n",
    "#save(interactions.all.res, file=paste0(output.dir, \"/\", run.id, \"/saved/interactions.all.res.Rdata\"))\n",
    "load(file=paste0(output.dir, \"/\", run.id, \"/saved/interactions.sig.res.Rdata\"))\n",
    "load(file=paste0(output.dir, \"/\", run.id, \"/saved/interactions.all.res.Rdata\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"running\"\n",
      "[1] \"Microglia\"         \"ExcitatoryNeurons\" \"MotorNeurons\"     \n"
     ]
    }
   ],
   "source": [
    "# Expand and annotate ATAC-seq peaks --------------------------------------\n",
    "\n",
    "print(\"running\")\n",
    "# Expand ATAC-seq peaks to minimum peak width and annotate with promoters/interactions.\n",
    "atac.seq.peaks.res <- list()\n",
    "atac.seq.peaks.ann <- list()\n",
    "print(cell.types)\n",
    "for (cell.type in cell.types) {\n",
    "  print(cell.type)\n",
    "  print(paste0(\"Expanding and annotating ATAC-seq peaks for cell type: \", cell.type))\n",
    "  atac.seq.peaks.res[[cell.type]] <- expandFeatures(atac.seq.peaks[[cell.type]], atac.seq.peak.width)\n",
    "  atac.seq.peaks.ann[[cell.type]] <- annotateATACSeqPeaks(atac.seq.peaks.res[[cell.type]], atac.seq.peaks[[cell.type]], interactions.sig.res[[cell.type]], promoters.strict, promoters.other, exons, introns)\n",
    "  \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save results.\n",
    "save(atac.seq.peaks.res, file=paste0(output.dir, \"/\", run.id, \"/saved/atac.seq.peaks.res.Rdata\"))\n",
    "save(atac.seq.peaks.ann, file=paste0(output.dir, \"/\", run.id, \"/saved/atac.seq.peaks.ann.Rdata\"))\n",
    "load(file=paste0(output.dir, \"/\", run.id, \"/saved/atac.seq.peaks.res.Rdata\"))\n",
    "load(file=paste0(output.dir, \"/\", run.id, \"/saved/atac.seq.peaks.ann.Rdata\"))\n",
    "\n",
    "# Plot/print ATAC-seq peak annotation results.\n",
    "# plotATACSeqPeakAnnotations(atac.seq.peaks.ann, paste0(output.dir, \"/\", run.id, \"/atac_seq_results\"))\n",
    "\n",
    "\n",
    "# Annotate interactions with ATAC-seq peaks and features ------------------\n",
    "\n",
    "\n",
    "# Annotate significant and nonsignificant interactions.\n",
    "interactions.sig.ann <- list()\n",
    "#interactions.low.ann <- list()\n",
    "for (cell.type in cell.types) {\n",
    "  \n",
    "  print(paste0(\"Annotating interactions for cell type: \", cell.type))\n",
    "  interactions.sig.ann[[cell.type]] <- annotateInteractions(interactions.sig.res[[cell.type]], atac.seq.peaks.res[[cell.type]], atac.seq.peaks.ann[[cell.type]], features[[cell.type]], promoters.strict, promoters.other)\n",
    "#  interactions.low.ann[[cell.type]] <- annotateInteractions(interactions.low.res[[cell.type]], atac.seq.peaks.res[[cell.type]], atac.seq.peaks.ann[[cell.type]], features[[cell.type]], promoters.strict, promoters.other)\n",
    "  \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'MotorNeurons'"
      ],
      "text/latex": [
       "'MotorNeurons'"
      ],
      "text/markdown": [
       "'MotorNeurons'"
      ],
      "text/plain": [
       "[1] \"MotorNeurons\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Save results.\n",
    "save(interactions.sig.ann, file=paste0(output.dir, \"/\", run.id, \"/saved/interactions.sig.ann.Rdata\"))\n",
    "#save(interactions.low.ann, file=paste0(output.dir, \"/\", run.id, \"/saved/interactions.low.ann.Rdata\"))\n",
    "load(file=paste0(output.dir, \"/\", run.id, \"/saved/interactions.sig.ann.Rdata\"))\n",
    "#load(file=paste0(output.dir, \"/\", run.id, \"/saved/interactions.low.ann.Rdata\"))\n",
    "\n",
    "# Plot interaction annotation results.\n",
    "plotInteractionAnnotations(interactions.sig.ann, paste0(output.dir, \"/\", run.id, \"/interaction_results/general\"))\n",
    "\n",
    "# Downsample nonsignificant interactions to same level as significant interactions.\n",
    "interactions.low.res.sampled <- list()\n",
    "interactions.low.ann.sampled <- list()\n",
    "for (cell.type in cell.types) {\n",
    "  \n",
    "  # Count # of significant interactions and sample random indices.\n",
    "  num.significant <- length(interactions.sig.res[[cell.type]][, 1])\n",
    "  num.nonsignificant <- length(interactions.low.res[[cell.type]][, 1])\n",
    "  sampled.indices <- sample(1:num.nonsignificant, num.significant, replace=F)\n",
    "  \n",
    "  # Sample nonsignificant interactions and their corresponding annotations.\n",
    "  interactions.low.res.sampled[[cell.type]] <- interactions.low.res[[cell.type]][sampled.indices, ]\n",
    "  interactions.low.ann.sampled[[cell.type]] <- interactions.low.ann[[cell.type]]\n",
    "  for (feature in names(interactions.low.ann[[cell.type]]))\n",
    "    interactions.low.ann.sampled[[cell.type]][[feature]] <- interactions.low.ann.sampled[[cell.type]][[feature]][sampled.indices]\n",
    "  \n",
    "}\n",
    "\n",
    "# Save results.\n",
    "save(interactions.low.res.sampled, file=paste0(output.dir, \"/\", run.id, \"/saved/interactions.low.res.sampled.Rdata\"))\n",
    "save(interactions.low.ann.sampled, file=paste0(output.dir, \"/\", run.id, \"/saved/interactions.low.ann.sampled.Rdata\"))\n",
    "load(file=paste0(output.dir, \"/\", run.id, \"/saved/interactions.low.res.sampled.Rdata\"))\n",
    "load(file=paste0(output.dir, \"/\", run.id, \"/saved/interactions.low.ann.sampled.Rdata\"))\n",
    "\n",
    "# Perform gene promoter hub analysis.\n",
    "hub.results.sig <- analyzeGenePromoterHubs(interactions.sig.res, interactions.sig.ann, promoters.strict, rpkm.data.individual)\n",
    "hub.results.low <- analyzeGenePromoterHubs(interactions.low.res.sampled, interactions.low.ann.sampled, promoters.strict, rpkm.data.individual)\n",
    "\n",
    "# Save results.\n",
    "save(hub.results.sig, file=paste0(output.dir, \"/\", run.id, \"/saved/hub.results.sig.Rdata\"))\n",
    "save(hub.results.low, file=paste0(output.dir, \"/\", run.id, \"/saved/hub.results.low.Rdata\"))\n",
    "load(file=paste0(output.dir, \"/\", run.id, \"/saved/hub.results.sig.Rdata\"))\n",
    "load(file=paste0(output.dir, \"/\", run.id, \"/saved/hub.results.low.Rdata\"))\n",
    "\n",
    "# Filter hub results for genes of interest. Plot gene promoter hubs histograms of # of interactions per gene.\n",
    "hub.results.sig.filtered <- filterHubResults(hub.results.sig, unique(promoters.strict$gene_id), -1, 1)\n",
    "plotGenePromoterHubsHistogram(hub.results.sig.filtered, cell.type.colors, paste0(output.dir, \"/\", run.id, \"/interaction_results/hubs\"), \"strict\")\n",
    "hub.results.sig.filtered <- filterHubResults(hub.results.sig, unique(promoters.strict$gene_id[promoters.strict$gene_type == \"protein_coding\"]), -1, 1)\n",
    "plotGenePromoterHubsHistogram(hub.results.sig.filtered, cell.type.colors, paste0(output.dir, \"/\", run.id, \"/interaction_results/hubs\"), \"protein_coding\")\n",
    "\n",
    "# Print # of interacting genes by cell type.\n",
    "printInteractingGenes(hub.results.sig, promoters.strict, cell.type.colors, paste0(output.dir, \"/\", run.id, \"/interaction_results/general\"))\n",
    "\n",
    "\n",
    "# Specificity analysis ----------------------------------------------------\n",
    "\n",
    "\n",
    "# Fill in heatmap of interaction scores at all significant interactions across all cell types.\n",
    "specificity.heatmap <- makeLociSpecificityHeatmap(interactions.sig.res, interactions.all.res, overlap.threshold)\n",
    "\n",
    "# Stitch together components of specificity matrix (generated using parallelized version of makeLociSpecificityHeatmap() function).\n",
    "load(file=paste0(output.dir, \"/\", run.id, \"/saved/specificity.heatmap.(\", cell.types[1], \").Rdata\"))\n",
    "specificity.matrix <- specificity.heatmap\n",
    "for (cell.type in cell.types) {\n",
    "  \n",
    "  load(file=paste0(output.dir, \"/\", run.id, \"/saved/specificity.heatmap.(\", cell.type, \").Rdata\"))\n",
    "  specificity.matrix[, cell.type] <- specificity.heatmap[, cell.type]\n",
    "  \n",
    "}\n",
    "head(specificity.matrix)\n",
    "specificity.table <- specificity.matrix >= score.cutoff\n",
    "\n",
    "# Make mapping of individual interaction IDs to cell type specificity.\n",
    "specificity.entries <- c()\n",
    "for (i in 1:length(specificity.table[, 1])) {\n",
    "  \n",
    "  interaction.ids <- unlist(strsplit(rownames(specificity.table)[i], split=\",\"))\n",
    "  for (interaction.id in interaction.ids)\n",
    "    specificity.entries <- rbind(specificity.entries, c(interaction.id, paste0(cell.types[specificity.table[i, ]], collapse=\",\")))\n",
    "  \n",
    "}\n",
    "specificity.entries[, 1] <- as.character(specificity.entries[, 1])\n",
    "specificity.entries[, 2] <- as.character(specificity.entries[, 2])\n",
    "\n",
    "# Save results.\n",
    "save(specificity.matrix, file=paste0(output.dir, \"/\", run.id, \"/saved/specificity.matrix.Rdata\"))\n",
    "save(specificity.table, file=paste0(output.dir, \"/\", run.id, \"/saved/specificity.table.Rdata\"))\n",
    "save(specificity.entries, file=paste0(output.dir, \"/\", run.id, \"/saved/specificity.entries.Rdata\"))\n",
    "load(file=paste0(output.dir, \"/\", run.id, \"/saved/specificity.matrix.Rdata\"))\n",
    "load(file=paste0(output.dir, \"/\", run.id, \"/saved/specificity.table.Rdata\"))\n",
    "load(file=paste0(output.dir, \"/\", run.id, \"/saved/specificity.entries.Rdata\"))\n",
    "\n",
    "# Plot specificity Venn diagrams.\n",
    "plotSpecificityVenn(specificity.table, cell.type.colors, paste0(output.dir, \"/\", run.id, \"/interaction_results/specificity\"))\n",
    "\n",
    "\n",
    "# GO analysis -------------------------------------------------------------\n",
    "\n",
    "\n",
    "# Set output prefix for writing input files for GO analysis.\n",
    "output.prefix <- paste0(output.dir, \"/\", run.id, \"/interaction_results/specificity\")\n",
    "\n",
    "# Find genes participating in interactions shared across all cell types.\n",
    "shared.terms <- getSpecificInteractingGenes(interactions.sig.res, interactions.sig.ann, atac.seq.peaks.ann, specificity.entries, c(\"cortical,hippocampal,motor,astrocyte\"), promoters.strict, rpkm.data.individual, 0.5, mode=\"both_atac\")\n",
    "write.table(shared.terms, file=paste0(output.prefix, \"/go.shared.txt\"), sep=\"\\t\", row.names=F, col.names=F, quote=F)\n",
    "\n",
    "# Find genes participating in specific interactions for each cell type.\n",
    "for (cell.type in cell.types) {\n",
    "  \n",
    "  specific.terms <- getSpecificInteractingGenes(interactions.sig.res, interactions.sig.ann, atac.seq.peaks.ann, specificity.entries, cell.type, promoters.strict, rpkm.data.individual, 0.5, mode=\"both_atac\")\n",
    "  specific.terms <- specific.terms[!(specific.terms %in% shared.terms)]\n",
    "  write.table(specific.terms, file=paste0(output.prefix, \"/go.\", cell.type, \".txt\"), sep=\"\\t\", row.names=F, col.names=F, quote=F)\n",
    "  \n",
    "}\n",
    "\n",
    "\n",
    "# Write interactions ------------------------------------------------------\n",
    "\n",
    "\n",
    "# Write interactions and their annotations to file.\n",
    "omit.list <- c(\"expanded_lhs\", \"expanded_lhs_ids\", \"expanded_rhs\", \"expanded_rhs_ids\")\n",
    "for (control.disease in c(\"AMD\", \"BH\", \"BMI\", \"RA\", \"T1D\", \"T2D\")) {\n",
    "  \n",
    "  omit.list <- c(omit.list, paste0(control.disease, \"_lhs\"))\n",
    "  omit.list <- c(omit.list, paste0(control.disease, \"_lhs_ids\"))\n",
    "  omit.list <- c(omit.list, paste0(control.disease, \"_rhs\"))\n",
    "  omit.list <- c(omit.list, paste0(control.disease, \"_rhs_ids\"))\n",
    "  \n",
    "}\n",
    "merged.data.all <- writeInteractions(interactions.sig, interactions.sig.ann, features, promoters.all, specificity.entries, omit.list, paste0(output.dir, \"/\", run.id, \"/interaction_results/general\"))\n",
    "\n",
    "# Save results.\n",
    "save(merged.data.all, file=paste0(output.dir, \"/\", run.id, \"/saved/merged.data.all.Rdata\"))\n",
    "load(file=paste0(output.dir, \"/\", run.id, \"/saved/merged.data.all.Rdata\"))\n",
    "\n",
    "\n",
    "# Vista element analysis --------------------------------------------------\n",
    "\n",
    "\n",
    "# Get list of unique annotations for Vista elements.\n",
    "all.annotations <- paste0(vista.data$annotation, collapse=\",\")\n",
    "all.annotations <- gsub(\"ganglion, cranial\", \"ganglion|cranial\", all.annotations)\n",
    "all.annotations <- unlist(strsplit(all.annotations, split=\",\"))\n",
    "for (i in 1:length(all.annotations))\n",
    "  all.annotations[i] <- unlist(strsplit(all.annotations[i], split=\"\\\\[\"))[1]\n",
    "unique.annotations <- unique(all.annotations)\n",
    "cns.annotations <- c(\"neural tube\", \"hindbrain (rhombencephalon)\", \"cranial nerve\", \"midbrain (mesencephalon)\", \"forebrain\", \n",
    "                     \"mesenchyme derived from neural crest\", \"dorsal root ganglion\", \"trigeminal V (ganglion|cranial)\")\n",
    "\n",
    "# Analyze overlap between Vista elements, interactions, and gene promoters.\n",
    "vista.results <- analyzeVistaOverlap(vista.data, interactions.sig.res, interactions.sig.ann, atac.seq.peaks.res, promoters.strict, promoters.all, rmap, interaction.resolution)\n",
    "\n",
    "# Save results.\n",
    "save(vista.results, file=paste0(output.dir, \"/\", run.id, \"/saved/vista.results.Rdata\"))\n",
    "load(file=paste0(output.dir, \"/\", run.id, \"/saved/vista.results.Rdata\"))\n",
    "\n",
    "# Print out results for Vista element analysis.\n",
    "plotVistaResults(vista.data, vista.results, cns.annotations, cell.type.colors, paste0(output.dir, \"/\", run.id, \"/feature_results/vista\"))\n",
    "printVistaConcordance(vista.results, cell.type.colors, paste0(output.dir, \"/\", run.id, \"/feature_results/vista\"))\n",
    "\n",
    "\n",
    "# GWAS SNP analysis -------------------------------------------------------\n",
    "\n",
    "\n",
    "# Perform overlap analysis for all SNPs.\n",
    "all.snp.results <- list()\n",
    "tag.snp.results <- list()\n",
    "for (disease in diseases) {\n",
    "  \n",
    "  print(paste0(\"Analyzing SNPs for disease: \", disease))\n",
    "  disease.data <- features[[cell.types[1]]][[\"snp\"]][[disease]]\n",
    "  all.snp.results[[disease]] <- analyzeAllSNPOverlap(disease.data, interactions.sig.res, interactions.sig.ann, atac.seq.peaks.res, atac.seq.peaks.ann,\n",
    "                                                     promoters.strict, promoters.all, rmap, 1000)\n",
    "  tag.snp.results[[disease]] <- analyzeTagSNPOverlap(disease.data, all.snp.results[[disease]], promoters.strict, promoters.all, rmap, 1000)\n",
    "  \n",
    "}\n",
    "\n",
    "# Save results.\n",
    "save(all.snp.results, file=paste0(output.dir, \"/\", run.id, \"/saved/all.snp.results.Rdata\"))\n",
    "save(tag.snp.results, file=paste0(output.dir, \"/\", run.id, \"/saved/tag.snp.results.Rdata\"))\n",
    "load(file=paste0(output.dir, \"/\", run.id, \"/saved/all.snp.results.Rdata\"))\n",
    "load(file=paste0(output.dir, \"/\", run.id, \"/saved/tag.snp.results.Rdata\"))\n",
    "\n",
    "# Print out results for SNP analysis.\n",
    "for (disease in diseases) {\n",
    "  \n",
    "  print(paste0(\"Processing results for disease: \", disease))\n",
    "  disease.data <- features[[cell.types[1]]][[\"snp\"]][[disease]]\n",
    "  plotSNPResults(disease.data, all.snp.results[[disease]], tag.snp.results[[disease]], cell.type.colors, disease, paste0(output.dir, \"/\", run.id, \"/feature_results/snps\"))\n",
    "  printSNPConcordance(tag.snp.results[[disease]])\n",
    "  getTargetGenes(all.snp.results[[disease]], promoters.all, rpkm.data.individual, 0, paste0(output.dir, \"/\", run.id, \"/feature_results/snps\"))\n",
    "  \n",
    "}\n",
    "\n",
    "# Print out results combined across all diseases.\n",
    "merged.results <- combineTagSNPResults(tag.snp.results, features, diseases)\n",
    "printSNPConcordance(merged.results)\n",
    "\n",
    "\n",
    "# Allelic bias annotation -------------------------------------------------\n",
    "\n",
    "\n",
    "# Annotate and write interactions with allelic bias for each cell type.\n",
    "allelic.interactions.sig.ann <- list()\n",
    "for (cell.type in c(\"cortical\", \"motor\")) {\n",
    "  \n",
    "  load(file=paste0(output.dir, \"/\", run.id, \"/saved/allelic.interactions.sig.pval.\", cell.type, \".Rdata\"))\n",
    "  print(length(allelic.interactions.sig[, 1]))\n",
    "  temp <- allelic.interactions.sig\n",
    "  allelic.interactions.sig <- list()\n",
    "  allelic.interactions.sig[[cell.type]] <- temp\n",
    "  allelic.interactions.sig.ann[[cell.type]] <- annotateInteractions(allelic.interactions.sig[[cell.type]], atac.seq.peaks.res[[cell.type]], atac.seq.peaks.ann[[cell.type]], features[[cell.type]], promoters.strict, promoters.other)\n",
    "  writeInteractions(allelic.interactions.sig, allelic.interactions.sig.ann, features, promoters.all, NULL, omit.list, paste0(output.dir, \"/\", run.id, \"/allelic_bias/results\"))\n",
    "  \n",
    "}\n",
    "\n",
    "\n",
    "# End ---------------------------------------------------------------------"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pchic_brain_R-3.6.2",
   "language": "R",
   "name": "r-3.6.2"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
